formats/xml/src/main/java/io/cloudevents/xml/XMLDeserializer.java

```diff
@@ -0,0 +1,338 @@
+     * in our context.
+     *
+     * @param e The root Element
+     * @throws CloudEventRWException
+     */
+    private void checkValidRootElement(Element e) throws CloudEventRWException {
+
+        // It must be the name we expect.
+        if (!XMLConstants.XML_ROOT_ELEMENT.equals(e.getLocalName())) {
+            throw CloudEventRWException.newInvalidDataType(e.getLocalName(), XMLConstants.XML_ROOT_ELEMENT);
+        }
+
+        // It must be in the CE namespace.
+        if (!XMLConstants.CE_NAMESPACE.equalsIgnoreCase(e.getNamespaceURI())) {
+            throw CloudEventRWException.newInvalidDataType(e.getNamespaceURI(), "Namespace: " + XMLConstants.CE_NAMESPACE);
+        }
+    }
+
+    /**
+     * Ensure the XML `data` element is well-formed.
+     *
+     * @param dataEl
+     * @throws CloudEventRWException
+     */
+    private void ensureValidDataElement(Element dataEl) throws CloudEventRWException {
+
+        // It must have a single child
+        final int childCount = XMLUtils.countOfChildElements(dataEl);
+        if (childCount != 1) {
+            throw CloudEventRWException.newInvalidDataType("data has " + childCount + " children", "1 expected");
+        }
+
+        // And there must be a valid type discriminator
+        final String xsiType = dataEl.getAttribute(XMLConstants.XSI_TYPE);
+
+        if (xsiType == null) {
+            throw CloudEventRWException.newInvalidDataType("NULL", "xsi:type oneOf [xs:base64Binary, xs:string, xs:any]");
+        }
+    }
+
+    /**
+     * Ensure a CloudEvent context attribute representation is as expected.
+     *
+     * @param el
+     * @throws CloudEventRWException
+     */
+    private void ensureValidContextAttribute(Element el) throws CloudEventRWException {
+
+        final String localName = el.getLocalName();
+
+        // It must be in our namespace
+        if (!XMLConstants.CE_NAMESPACE.equals(el.getNamespaceURI())) {
+            final String allowedTxt = el.getLocalName() + " Expected namespace: " + XMLConstants.CE_NAMESPACE;
+            throw CloudEventRWException.newInvalidDataType(el.getNamespaceURI(), allowedTxt);
+        }
+
+        // It must be all lowercase
+        if (!allLowerCase(localName)) {
+            throw CloudEventRWException.newInvalidDataType(localName, " context atttribute names MUST be lowercase");
+        }
+
+        // A bit of a kludge, not relevant for 'data' - should refactor
+        if (!XMLConstants.XML_DATA_ELEMENT.equals(localName)) {
+            // It must not have any children
+            if (XMLUtils.countOfChildElements(el) != 0) {
+                throw CloudEventRWException.newInvalidDataType(el.getLocalName(), "Unexpected child element(s)");
+            }
+        }
+
+        // Finally, ensure we only see each CE Attribute once...
+        try {
+            ceAttributeTracker.trackOccurrence(localName);
+        } catch (IllegalStateException e){
+            throw CloudEventRWException.newOther(e);
+        }
```

Comment:
Would a Map-like API work here, instead of an exception-based control flow?
```java
if (ceAttributeTracker.trackOccurrence(localName)) {
   throw CloudEventRWException.new...
}
```

Reply:
I will make this change ...